crypto_sign/tuov_ip/ref/parallel_matrix_op.c

```diff
@@ -0,0 +1,564 @@
+            gf256v_add( bC, tmp_c, size_batch);
+            bC += size_batch;
+        }
+        btriA += size_batch * (Aheight - i - 1);
+    }
+}
+
+// This function is only used in calssic mode.
+void batch_trimatTr_madd_gf16( unsigned char *bC, const unsigned char *btriA,
+                               const unsigned char *B, unsigned Bheight, unsigned size_Bcolvec, unsigned Bwidth, unsigned size_batch ) {
+#define MAX_O_BYTE  (32)
+#define MAX_V      (96)
+    uint8_t tmp_c[MAX_O_BYTE];
+    uint8_t tmp_Arow[MAX_V * MAX_O_BYTE];
+#undef MAX_O_BYTE
+#undef MAX_V
+
+// access fixed positions of destination matrix C
+    unsigned Aheight = Bheight;
+    for (unsigned i = 0; i < Aheight; i++) {
+        const uint8_t *ptr = btriA + i * size_batch;
+        for (unsigned j = 0; j < i; j++) {
+            memcpy( tmp_Arow + j * size_batch, ptr, size_batch );
+            ptr += (Aheight - j - 1) * size_batch;
+        }
+        memcpy( tmp_Arow + i * size_batch, ptr, size_batch );
+
+        for (unsigned j = 0; j < Bwidth; j++) {
+            gf16mat_prod( tmp_c, tmp_Arow, size_batch, i + 1, B + (j * size_Bcolvec) );
+            gf256v_add( bC, tmp_c, size_batch);
+            bC += size_batch;
+        }
+    }
+}
+
+void batch_2trimat_madd_gf16( unsigned char *bC, const unsigned char *btriA,
+                              const unsigned char *B, unsigned Bheight, unsigned size_Bcolvec, unsigned Bwidth, unsigned size_batch ) {
+#define MAX_O_BYTE  (32)
+#define MAX_V      (96)
+    uint8_t tmp_c[MAX_O_BYTE];
+    uint8_t tmp_Arow[MAX_V * MAX_O_BYTE];
+#undef MAX_O_BYTE
+#undef MAX_V
+
+// access fixed positions of destination matrix C
+    unsigned Aheight = Bheight;
+    for (unsigned i = 0; i < Aheight; i++) {
+        const uint8_t *ptr = btriA + i * size_batch;
+        for (unsigned j = 0; j < i; j++) {
+            memcpy( tmp_Arow + j * size_batch, ptr, size_batch );
+            ptr += (Aheight - j - 1) * size_batch;
+        }
+        memset( tmp_Arow + i * size_batch, 0, size_batch );
+        ptr += size_batch;
+        memcpy( tmp_Arow + (i + 1)*size_batch, ptr, size_batch * (Aheight - i - 1) );
+
+        for (unsigned j = 0; j < Bwidth; j++) {
+            gf16mat_prod( tmp_c, tmp_Arow, size_batch, Aheight, B + (j * size_Bcolvec) );
+            gf256v_add( bC, tmp_c, size_batch);
+            bC += size_batch;
+        }
+    }
+}
+
+////////////////
+
+
+
+void batch_upper_matTr_x_mat_gf16( unsigned char *bC, const unsigned char *A_to_tr, unsigned Aheight, unsigned size_Acolvec, unsigned Awidth,
+                                   const unsigned char *bB, unsigned Bwidth, unsigned size_batch ) {
+#define MAX_O  (64)
+#define MAX_O_BYTE  (32)
+    uint8_t _ALIGN_(32) row[MAX_O * MAX_O_BYTE]; /// XXX: buffer for maximum row
+#undef MAX_O_BYTE
+#undef MAX_O
+    unsigned Atr_height = Awidth;
+    unsigned Atr_width  = Aheight;
+    uint8_t *A_copy;
+    if (size_Acolvec == 1){
+        A_copy = malloc(Aheight * Awidth);
```

Comment:
The only other `malloc`/`free` I could find. Can be easily replaced by a VLA. Again, no idea about the size.